### DISCUSSION TITLE
example trainer compatible with and without BatchNorm

### DISCUSSION THREAD

**Discussion by luweizheng** [google/flax#2189] [Q&A] [answered]
Hi all,

I am working on a framework based on jax and flax. As jax is fully functional, we should manage state. 
What I want to do is designing a trainer. Users pass their own `nn.Module` into the trainer.
For BatchNorm, flax has specific `batch_stats` to manage state. So If a user's `nn.Module` contains `BatchNorm`, the trainer have to use a different way than without `BatchNorm`. 

Is there any trainer example code for me to reference? Maybe HuggingFace?

**Comment by marcvanzee** [comment]
T5X has a trainer abstraction but it fairly complex: https://github.com/google-research/t5x/blob/main/t5x/trainer.py

HuggingFace doesn't provide one, see here: https://github.com/huggingface/transformers/blob/main/examples/flax/README.md

Elegy provides a high-level API for Haiku and Flax and has a `fit` method: https://github.com/poets-ai/elegy

@cgarciae maybe knows more?

**Reply by luweizheng** [reply]
Thanks a lot!

**Comment by cgarciae** [comment, marked as answer]
Hey @luweizheng! Elegy is a trainer interface for JAX that supports Flax in case you want to leverage it:

https://github.com/poets-ai/elegy

If instead you need a reference implementation on how setup a generic Flax workflow maybe check how Elegy implements support for Flax:

https://github.com/poets-ai/elegy/blob/vnext/elegy/modules/high_level/flax_module.py#L134-L210

Note: this code is part of Elegy's [vnext](https://github.com/poets-ai/elegy/tree/vnext) branch which is still in alpha but its easier to understand than how the main branch handles it, just beware it might contain errors.

**Reply by luweizheng** [reply]
Thanks a lot!
